torch/_refs/__init__.py

```diff
@@ -6221,10 +6233,6 @@ def numel_error():
 
 @register_decomposition(aten.dot)
 @out_wrapper()
-@elementwise_type_promotion_wrapper(
-    type_promoting_args=("self", "other"),
-    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
-)
 def dot(self, other):
     if self.is_complex():
         if self.is_conj():
```

Comment:
I did it this way because the complex branch is effectively the same implementation as the C++ implementation (e.g. for dot CPU: https://github.com/pytorch/pytorch/blob/392221b39036bb140e32f428fc6e318744005651/aten/src/ATen/native/Blas.cpp#L159). So the computation and result dtype in this branch should be correct, as the torch.dot / vdot would do this if necessary.

The actual _decomposition_ into an elementwise product followed by a reduction is only for the case when the inputs are real, so type promotion is applied.

That being said, if the computation and result types for dot, vdot and sum + reduction should be the same then it would be cleaner to do what you suggest - I just didn't want to make the assumption that they are.

Reply:
Looks like `dot_naive` use the same type promotion rules anyway: https://github.com/pytorch/pytorch/blob/392221b39036bb140e32f428fc6e318744005651/aten/src/ATen/native/Blas.cpp#L159.

I'll change the code to your suggestion